src/libstd/sys/windows/backtrace/printing/msvc.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright 2014 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use borrow::Cow;
+use ffi::CStr;
+use io::prelude::*;
+use io;
+use libc::{c_ulong, c_int, c_char, c_void};
+use mem;
+use sys::c;
+use sys::dynamic_lib::DynamicLibrary;
+use sys::backtrace::BacktraceContext;
+use sys_common::backtrace::{output, output_fileline};
+
+use super::super::LOCK;
+
+type SymFromAddrFn =
+    unsafe extern "system" fn(c::HANDLE, u64, *mut u64,
+                              *mut c::SYMBOL_INFO) -> c::BOOL;
+type SymGetLineFromAddr64Fn =
+    unsafe extern "system" fn(c::HANDLE, u64, *mut u32,
+                              *mut c::IMAGEHLP_LINE64) -> c::BOOL;
+
+/// Converts a pointer to symbol to its string value.
+pub fn resolve_symname(symbol_addr: Frame, context: &BacktraceContext)
+    -> Option<Cow<'static, str>>
+{
+    unsafe {
+        LOCK.lock();
+        let res = _resolve_symname(symbol_addr, context);
+        LOCK.unlock();
+        res
+    }
+}
+
+fn _resolve_symname(symbol_addr: Frame, context: &BacktraceContext)
+    -> Option<Cow<'static, str>>
+{
+    unsafe {
+        let SymFromAddr = sym!(&context.dbghelp, "SymFromAddr", SymFromAddrFn);
+
+        let mut info: c::SYMBOL_INFO = mem::zeroed();
+        info.MaxNameLen = c::MAX_SYM_NAME as c_ulong;
+        // the struct size in C.  the value is different to
+        // `size_of::<SYMBOL_INFO>() - MAX_SYM_NAME + 1` (== 81)
+        // due to struct alignment.
+        info.SizeOfStruct = 88;
+
+        let mut displacement = 0u64;
+        let ret = SymFromAddr(context.process,
+                              symbol_addr as u64,
+                              &mut displacement,
+                              &mut info);
+
+        if ret == c::TRUE {
+            let ptr = info.Name.as_ptr() as *const c_char;
+            CStr::from_ptr(ptr).to_str().ok().map(|s| Cow::Owned(s.into()))
```

Comment:
I think this is still an issue in the sense that we should use a closure here. We should strive to avoid allocations in backtraces because they're often generated in limited context situations (panics, aborts, etc), where touching as little as possible is advantageous.